### ISSUE TITLE
Inconsistent outputs of tcpdump

### ISSUE THREAD

**Issue by shao-hua-li** [the-tcpdump-group/tcpdump#919]
Hi there,

I found that `tcpdump` compiled with gcc11 and clang12 would give different results for some inputs. I have attached a file that could reveal this issue [diff.pcap](https://drive.google.com/file/d/1Nh8BIH4_06ets9gHJQbXoeqpq4_CmstM/view?usp=sharing)

- tcpdump version: `4.99.0`
- libpcap version: `1.10.0`
- platform: `Ubuntu 20.04.2 LTS, x86-64`
- compilers: `gcc11, clang12`
- Reproduce: run `tcpdump -e -vv -nr diff.pcap` with different compilers's tcpdump. The `Source-AS` field for gcc11 and clang12 compiled tcpdump would be different. In my machine, gcc11 was `Source-AS 2326532457`, clang12 was `Source-AS 4215040`.

**Comment by gvanem** [comment]
Your `diff.pcap` is bogus:
```
windump.exe: pcap_loop: invalid packet capture length 4294967295, bigger than snaplen of 65535
```

**Comment by fxlb** [comment]
Thank you for the report. Problem reproduced. Fix ongoing ...

**Comment by fxlb** [comment]
Should be fixed by cf6c3fb2ad11e3ce94401b4371d91def79934310 in master branch and 8ce040a39025e35864973630ff2006f2313d51cf in tcpdump-4.99 branch.

**Comment by shao-hua-li** [comment]
I checked that this issue has been solved in the head versions of tcpdump and libpcap. Thanks :).

**Comment by shao-hua-li** [comment]
Hmm, I found another similar issue in `print-arp.c`. The two `tcpdump -e -vv -nr` compiled with gcc11 and clang12 would again output differently for the file [diff_2.pcap](https://drive.google.com/file/d/1DP5J8mAO4tV8wQEBPlkbHHgvBVSqbwfl/view?usp=sharing).

The `who-is` and `tell` for gcc11 would be `DLCI 3776`, while blank for clang12.

**Comment by fxlb** [comment]
Shoud be another issue.

**Comment by guyharris** [comment]
> Should be another issue.

That depends on whether the issue is viewed as "the BGP dissector produces different results with different compilers" or "various dissectors produce different results with different compilers".

The underlying problem in both cases is that if you have a routine that converts something to a string by formatting the string into a static buffer and returning a pointer to the buffer, it is *NOT* safe to use that routine more than once as an argument to a function, as the first call that is performed will have its result overwritten by the second call that is performed, and the same result pointer will be passed more than once to the function being called.

I.e., the result will be wrong no matter in *what* order the two calls to the function using the static buffer are made, but the form of "wrong" will differ based on the order in which they're done; the code generated by different compilers may evaluate the arguments to the function in different orders.

The offending static-buffer function in the ARP case is `q922_string()`.

Wireshark originally handled this by having some functions have an array of a small number of static buffers, cycling through them, so as long as you made fewer than number-of-static-buffers calls within a function call, you were OK.

This is all because C's handling of array data values, including string data values, is a bit weak.  C++ can handle this a little better with classes such as the `String` class, where the destructor can be called when the value is no longer accessible.  Thus, C programmers tend to use either hacks such as "return a pointer to a static buffer", or have functions that take a buffer as an argument.  The former has the obvious problem (plus issues in a threaded environment); the latter is a bit of a pain as you have to declare multiple local buffers.

A while ago, Wireshark added a "poor man's generational garbage collection" scheme, in which there's a memory allocation pool that, when dissection is complete, gets freed in bulk, with routines returning pointers to strings allocated from that pool.  (It now has multiple such pools with different lifetimes.)

For now, we should probably just have `linkaddr_string()`, and the code that calls it, take a buffer pointer, into which a string will be formatted if necessary.  I'll look at that.

A cleaner longer-term mechanism might be a printf-like routine that allocates a sufficiently-large buffer for the string to be formatted (like `asprintf()`), ensure that this buffer will be freed once dissection is complete, and return a pointer to that buffer, and use that for formatting, Wireshark-style.

**Comment by guyharris** [comment]
I decided that the quick fix is just to split a couple of 'who-is XXX tell YYY" printouts into two calls, so there's only one `GET_LINKADDR_STRING()` call per `ND_PRINT()` call - it's less disruptive.

We should consider the cleaner long-term mechanism for 5.0 or later, as that gets rid of static buffers, means you don't have to manage auto buffers, and means you don't have to pick an appropriate size for those buffers as they're allocated to be big enough for whatever string is being formatted.

I'm also seeing differences between tcpdump's and Wireshark's dissection of packet 2, which may involve fixes to both, but I'll do that later.

**Comment by guyharris** [comment]
> I'm also seeing differences between tcpdump's and Wireshark's dissection of packet 2

One difference is that Wireshark doesn't bother dissecting ARP hardware addresses that aren't Ethernet-like, so it just shows raw hex digits rather than a DLCI number for Q.922 addresses.

The other difference is due to tcpdump's `q922_string()` not filling in the static string buffer if `parse_q922_header()` fails; the Q.922 addresses in that ARP packet are too long (probably because the packet was fuzzed and the hardware address type was changed to something non-MAC-address-like).  That needs to be fixed in tcpdump; I'll do that.

**Comment by guyharris** [comment]
> That needs to be fixed in tcpdump; I'll do that.

Done in 51bc4bffb849bd323fcff9a03b8b5492885fe37a in main and in a7dcc3c1779a63517f23d63f059287cfae2caf3e in the 4.99 branch.

**Comment by guyharris** [comment]
> Your diff.pcap is bogus:

There really needs to be a guide to how to fuzz packet captures.

One very important rule:

If you're testing tcpdump or Wireshark, and possibly if you're testing other sniffers, the software you're testing has low-level code that understands the format of capture files but doesn't understand anything about packet data and higher-level code that doesn't understand the format of capture files but does understand packet data.

(The libraries that reads the capture files - libpcap for tcpdump, libwiretap for Wireshark - have a small amount of code that does understand a little bit about packet data, but there's still a separation between the two bits of code.)

Capture files generally have:

- some non-per-packet data (file header in pcap, non-packet blocks in pcapng, etc.);
- packets, which themselves have:

- a header, and possibly trailer, for each packet;
- the raw packet data.

If you want to test the low-level code, fuzz the non-per-packet data and the per-packet header and, if present, trailer; you can fuzz the packet data, but if you fuzz the header or trailer, you may run the risk of the code *not* being buggy, detecting the damage, and not ready any further into the file, so anything you've fuzzed in the packet data past that point will not be tested.  That's what happened here - anything fuzzed after the first packet in diff.pcap, or after the fourth packet in diff_2.pcap, is *not* being tested, because libpcap gives up on the file with the second packet in diff.pcap and the fifth packet in diff_2.pcap.

If you want to test the high-level code, leave *everything* except for the packet data alone, so that the low-level code doesn't see a problem and quit reading the file.

**Comment by shao-hua-li** [comment]
Thanks for your advice, that makes a lot of sense :-D.
